plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,102 +64,126 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
+    }
+
+    /**
+     * IN restriction allowed only on last clustering column for Cassandra version <= 2.1
+     */
+    private static boolean isInStatementNotAllowed(List<CassandraColumnHandle> clusteringColumns, VersionNumber cassandraVersion, int currentlyProcessedClusteringColumn)
+    {
+        return cassandraVersion.compareTo(VersionNumber.parse("2.2.0")) < 0 && currentlyProcessedClusteringColumn != (clusteringColumns.size() - 1);
+    }
+
+    private static String translateRangeIntoCql(CassandraColumnHandle columnHandle, Range range)
+    {
+        if (range.isSingleValue()) {
+            return format("%s = %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    columnHandle.getCassandraType().toCqlLiteral(range.getSingleValue()));
+        }
+
+        String lowerBoundPredicate = null;
+        String upperBoundPredicate = null;
+        if (!range.isLowUnbounded()) {
+            String lowBound = columnHandle.getCassandraType().toCqlLiteral(range.getLowBoundedValue());
+            lowerBoundPredicate = format(
+                    "%s %s %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    range.isLowInclusive() ? ">=" : ">",
+                    lowBound);
+        }
+        if (!range.isHighUnbounded()) {
+            String highBound = columnHandle.getCassandraType().toCqlLiteral(range.getHighBoundedValue());
+            upperBoundPredicate = format(
+                    "%s %s %s",
+                    CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                    range.isHighInclusive() ? "<=" : "<",
+                    highBound);
+        }
+        if (lowerBoundPredicate != null && upperBoundPredicate != null) {
+            return format("%s AND %s ", lowerBoundPredicate, upperBoundPredicate);
+        }
+        if (lowerBoundPredicate != null) {
+            return lowerBoundPredicate;
+        }
+        return upperBoundPredicate;
     }
 
     private static class ClusteringPushDownResult
     {
-        private final Map<ColumnHandle, Domain> domains;
+        private final Set<ColumnHandle> fullyPushedColumns;
         private final String domainQuery;
 
-        public ClusteringPushDownResult(Map<ColumnHandle, Domain> domains, String domainQuery)
+        public ClusteringPushDownResult(Set<ColumnHandle> columnPredicates, String domainQuery)
         {
-            this.domains = requireNonNull(ImmutableMap.copyOf(domains));
-            this.domainQuery = requireNonNull(domainQuery);
+            this.fullyPushedColumns = ImmutableSet.copyOf(requireNonNull(columnPredicates, "columnPredicates is null"));
+            this.domainQuery = requireNonNull(domainQuery, "domainQuery is null");
         }
 
-        public Map<ColumnHandle, Domain> getDomains()
+        public boolean hasNotBeenFullyPushed(ColumnHandle column)
```

Comment:
i would rather let the calling code to apply negation, so that method semantics are clearer.